enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,328 @@
+resource.
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.
+
+How will security be reviewed and by whom? How will UX be reviewed and by whom?
+
+Consider including folks that also work outside your immediate sub-project.
+
+## Design Details
+
+### Open Questions
+
+1.
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. Initial proposal
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this
+enhancement:
+
+- Maturity levels
+    - [`alpha`, `beta`, `stable` in upstream Kubernetes][maturity-levels]
+    - `Dev Preview`, `Tech Preview`, `GA` in OpenShift
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning),
+or by redefining what graduation means.
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+##### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this
```

Comment:
This seems tricky.  What happens if we grow a new HA enum (say), and the user updates to it before all cluster components understand that value?  Looks like the config operator is currently [at level 30][1], so before the bulk of operator at level 50, but after etcd ([at level 12][2]) and other early operators.  That means that a config operator would learn of the new HA enum early, and could accept the user setting it before the update completed and later operators understood what to do about the new enum.  If we never update in any of the non-HA modes, that's ok.  But the same issues would apply to `resourceLimits` if that's a `spec`-configurable knob.  Maybe components could scrape `Capabilities`'s `status` and reflect their currently-accepted value in their own `status` (on their ClusterOperator or a per-component config object or whatever).  That way they could keep rolling forward with whatever their previously accepted config was if the central value was unrecognized, hoping that soon they would be rebooted into a new version which understood the newly-accepted central value.

[1]: https://github.com/openshift/cluster-config-operator/tree/351f7aa07c0886866484f9916ebb2b2c5663c0e1/manifests
[2]: https://github.com/openshift/cluster-etcd-operator/tree/06667d94d6313d6edd59801a351de7038987f36d/manifests

Reply:
Maybe that's another argument for making the API immutable, at least to start?

Reply:
Thinking about this more, it seems error prone to put logic to manage this in every operator that consumes the API. I've captured that idea in the alternatives, and proposed a way to manage the settings in the new API to avoid the duplication of effort.